File: src/lib/ical.ts
import { env } from "./env";

/**
 * Minimal RFC-5545 iCalendar parser — enough to pull the busy/blocked date
 * ranges out of an OTA export feed (Airbnb / Booking / Vrbo all publish these).
 * No external dependency.
 */
export type IcalEvent = {
  uid: string | null;
  summary: string | null;
  start: string | null; // YYYY-MM-DD
  end: string | null; // YYYY-MM-DD (exclusive, per iCal DTEND for all-day)
  status: string | null;
  source: "reserved" | "blocked" | "other";
};

export type IcalFeed = {
  url: string;
  fetched_at: string;
  calendar_name: string | null;
  event_count: number;
  events: IcalEvent[];
  blocked_dates: string[]; // flat sorted list of every YYYY-MM-DD that is unavailable
  error: string | null;
};

function unfold(raw: string): string[] {
  // Continuation lines start with a space or tab.
  return raw
    .replace(/\r\n/g, "\n")
    .split("\n")
    .reduce<string[]>((acc, line) => {
      if (/^[ \t]/.test(line) && acc.length) acc[acc.length - 1] += line.slice(1);
      else acc.push(line);
      return acc;
    }, []);
}

function parseIcalDate(v: string): string | null {
  const m = v.match(/(\d{4})(\d{2})(\d{2})/);
  return m ? `${m[1]}-${m[2]}-${m[3]}` : null;
}

function eachDate(start: string, endExclusive: string): string[] {
  const out: string[] = [];
  const d = new Date(start + "T00:00:00Z");
  const end = new Date(endExclusive + "T00:00:00Z");
  let guard = 0;
  while (d < end && guard++ < 1000) {
    out.push(d.toISOString().slice(0, 10));
    d.setUTCDate(d.getUTCDate() + 1);
  }
  return out;
}

export function parseIcal(text: string): Omit<IcalFeed, "url" | "fetched_at" | "error"> {
  const lines = unfold(text);
  let calendarName: string | null = null;
  const events: IcalEvent[] = [];
  let cur: Partial<IcalEvent> & { _raw?: Record<string, string> } | null = null;

  for (const line of lines) {
    const idx = line.indexOf(":");
    if (idx === -1) continue;
    const keyPart = line.slice(0, idx);
    const value = line.slice(idx + 1).trim();
    const key = keyPart.split(";")[0].toUpperCase();

    if (key === "X-WR-CALNAME") calendarName = value;
    else if (line.toUpperCase().startsWith("BEGIN:VEVENT")) cur = {};
    else if (line.toUpperCase().startsWith("END:VEVENT")) {
      if (cur) {
        const summary = cur.summary ?? null;
        const status = cur.status ?? null;
        const s = /reserv|booked|guest/i.test(`${summary} ${status}`)
          ? "reserved"
          : /not available|blocked|unavailable|closed|owner/i.test(`${summary} ${status}`)
            ? "blocked"
            : "other";
        events.push({
          uid: cur.uid ?? null,
          summary,
          start: cur.start ?? null,
          end: cur.end ?? null,
          status,
          source: s as IcalEvent["source"],
        });
      }
      cur = null;
    } else if (cur) {
      if (key === "UID") cur.uid = value;
      else if (key === "SUMMARY") cur.summary = value;
      else if (key === "STATUS") cur.status = value;
      else if (key === "DTSTART") cur.start = parseIcalDate(value);
      else if (key === "DTEND") cur.end = parseIcalDate(value);
    }
  }

  const blocked = new Set<string>();
  for (const e of events) {
    if (!e.start) continue;
    const end = e.end ?? e.start;
    for (const d of eachDate(e.start, end)) blocked.add(d);
  }

  return {
    calendar_name: calendarName,
    event_count: events.length,
    events,
    blocked_dates: [...blocked].sort(),
  };
}

export async function fetchIcal(url: string): Promise<IcalFeed> {
  const base: IcalFeed = {
    url,
    fetched_at: new Date().toISOString(),
    calendar_name: null,
    event_count: 0,
    events: [],
    blocked_dates: [],
    error: null,
  };
  let u: URL;
  try {
    u = new URL(url);
  } catch {
    return { ...base, error: "invalid iCal URL" };
  }
  if (!/^https?:$/.test(u.protocol)) return { ...base, error: "URL must be http(s)" };

  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), env.fetchTimeoutMs);
    const res = await fetch(u.toString(), {
      signal: ctrl.signal,
      headers: { "user-agent": "HostiggoImporter/1.0", accept: "text/calendar,*/*" },
    }).finally(() => clearTimeout(t));
    if (!res.ok) return { ...base, error: `HTTP ${res.status}` };
    const text = await res.text();
    if (!/BEGIN:VCALENDAR/i.test(text))
      return { ...base, error: "response is not an iCalendar feed" };
    return { ...base, ...parseIcal(text) };
  } catch (e) {
    return { ...base, error: (e as Error).message };
  }
}

File: service/app/pipeline/ical.py
from __future__ import annotations

import re
from datetime import UTC, date, datetime, timedelta

import httpx

from app.config import settings
from app.models import IcalEvent, IcalFeed

_RESERVED = re.compile(r"reserv|booked|guest", re.IGNORECASE)
_BLOCKED = re.compile(r"not available|blocked|unavailable|closed|owner", re.IGNORECASE)


def _unfold(raw: str) -> list[str]:
    out: list[str] = []
    for line in raw.replace("\r\n", "\n").split("\n"):
        if line[:1] in (" ", "\t") and out:
            out[-1] += line[1:]
        else:
            out.append(line)
    return out


def _parse_date(v: str) -> str | None:
    m = re.search(r"(\d{4})(\d{2})(\d{2})", v)
    return f"{m.group(1)}-{m.group(2)}-{m.group(3)}" if m else None


def _each_date(start: str, end_exclusive: str) -> list[str]:
    d = date.fromisoformat(start)
    end = date.fromisoformat(end_exclusive)
    out: list[str] = []
    while d < end and len(out) < 1000:
        out.append(d.isoformat())
        d += timedelta(days=1)
    return out


def parse_ical(text: str) -> dict:
    lines = _unfold(text)
    calendar_name: str | None = None
    events: list[IcalEvent] = []
    cur: dict | None = None

    for line in lines:
        if ":" not in line:
            continue
        key_part, _, value = line.partition(":")
        value = value.strip()
        key = key_part.split(";")[0].upper()

        if key == "X-WR-CALNAME":
            calendar_name = value
        elif line.upper().startswith("BEGIN:VEVENT"):
            cur = {}
        elif line.upper().startswith("END:VEVENT"):
            if cur is not None:
                blob = f"{cur.get('summary')} {cur.get('status')}"
                kind = (
                    "reserved" if _RESERVED.search(blob)
                    else "blocked" if _BLOCKED.search(blob)
                    else "other"
                )
                events.append(IcalEvent(
                    uid=cur.get("uid"), summary=cur.get("summary"),
                    start=cur.get("start"), end=cur.get("end"),
                    status=cur.get("status"), kind=kind,
                ))
            cur = None
        elif cur is not None:
            if key == "UID":
                cur["uid"] = value
            elif key == "SUMMARY":
                cur["summary"] = value
            elif key == "STATUS":
                cur["status"] = value
            elif key == "DTSTART":
                cur["start"] = _parse_date(value)
            elif key == "DTEND":
                cur["end"] = _parse_date(value)

    blocked: set[str] = set()
    for e in events:
        if not e.start:
            continue
        for d in _each_date(e.start, e.end or e.start):
            blocked.add(d)

    return {
        "calendar_name": calendar_name,
        "event_count": len(events),
        "events": events,
        "blocked_dates": sorted(blocked),
    }


async def fetch_ical(url: str) -> IcalFeed:
    base = IcalFeed(url=url, fetched_at=datetime.now(UTC).isoformat())
    if not re.match(r"^https?://", url):
        base.error = "URL must be http(s)"
        return base
    try:
        async with httpx.AsyncClient(timeout=settings.import_fetch_timeout_s) as c:
            r = await c.get(url, headers={
                "user-agent": "HostiggoImporter/1.0", "accept": "text/calendar,*/*"
            })
        if r.status_code != 200:
            base.error = f"HTTP {r.status_code}"
            return base
        if "BEGIN:VCALENDAR" not in r.text:
            base.error = "response is not an iCalendar feed"
            return base
        parsed = parse_ical(r.text)
        return base.model_copy(update=parsed)
    except httpx.HTTPError as e:
        base.error = str(e)
        return base
